### ISSUE TITLE
Why is PNG compression poor?

### ISSUE THREAD

**Issue by 9mm** [ImageOptim/ImageOptim#230]
I've honestly searched far and wide for good compression for both PNG and JPG. In my experience it's been an absolute nightmare getting both in the same place, let alone separately half the time.

I was really pleased with ImageOptim with the JPG compression, but in most PNG's I try to compress it's 30-40% worse than other tools I use.

A great baseline that I use is http://compresspng.com/ and second is https://tinypng.com/

Both of these sites get PNG's so much smaller.

I've had images where both sites above get ~75% PNG decrease but imageOptim is 3%. I've tried tons of other PNG's and if the average is ~65% on the 2 sites above it will be like.... 15% on ImageOptin.

And yes, I do have lossy option checked, but I had it set to ~70-80%. It honestly doesnt make a difference from what I really see though, because even when its at 60% it will won't change the file size by more than 1-2% (for the images i tried)

How are those 2 sites so good at compressing PNGs? I would LOVE to have all my compression in one place, your software is so amazing otherwise.

**Comment by pornel** [comment]
I wrote engine that powers both compresspng and tinypng. 

You can get the same or better if you enable Lossy option in ImageOptim.

**Comment by 9mm** [comment]
Wow, the real OG 🙏🏼🙏🏼

I have enabled lossy, but it doesn't even come close to tinypng.com. I haven't tried a SUPER low quality like 20% though...

What % quality does tinypng use?

**Comment by 9mm** [comment]
oops

**Comment by 9mm** [comment]
Here's an example for 3 demo PNG's from a random project.

This is with ImageOptim set to <strong>40% quality</strong>! (as low as it goes)

<img width="703" alt="screen shot 2017-08-16 at 9 55 36 am" src="https://user-images.githubusercontent.com/1386207/29366861-26f17a3e-8269-11e7-8840-a7550507a28b.png">

Here's TinyPNG:

<img width="1000" alt="screen shot 2017-08-16 at 9 56 27 am" src="https://user-images.githubusercontent.com/1386207/29366887-34a80788-8269-11e7-9e60-57cf5fd89d76.png">

And here' is my favorite one, which almost always beats everything I ever try... CompressPNG:

<img width="561" alt="screen shot 2017-08-16 at 9 57 06 am" src="https://user-images.githubusercontent.com/1386207/29366931-4c92a308-8269-11e7-9e69-f8afab5dd94b.png">

**Comment by 9mm** [comment]
Here was the images I used

[images.zip](https://github.com/ImageOptim/ImageOptim/files/1228424/images.zip)

**Comment by pornel** [comment]
All of these tools use pngquant for the majority of filesize reduction. 

The difference is tinypng reduces number of colors in the image more aggressively, risking quality degradation.

In ImageOptim I reduce number of colors only if it reduction in file size is bigger than reduction in quality. Even when you set quality to 40%, ImageOptim treats it as lowest acceptable quality, but it will use higher quality if it gives better "bang for the buck".

The logo from TinyPNG has 96 colors. The logo from ImageOptim has 196 colors — more than twice as much for 0.3KB in size.

These are quite small and simple images, so the difference is not as visible, but in more colorful images that becomes important.

**Comment by Saklad5** [comment]
Is there a reason this issue is still open?

**Comment by pornel** [comment]
I'm going to make low PNG quality reduce the number of colors below 256, to match the other compressors.

**Comment by 9mm** [comment]
Awesome!

**Comment by Saklad5** [comment]
ImageOptim is designed to be a GUI for lossless compression tools. Unless you want to expand that description, I wouldn’t feel too pressured to prioritize file size over quality. Most people who are looking for that would use something else, right?

**Comment by 9mm** [comment]
That seems like quite a stretch... some people want lower file sizes and they're OK with small reductions in quality (i.e. pretty much every mobile web PNG I've ever tried to compress).

If people wanted to use a different tool for everything that defeats the purpose of a nice app like ImageOptim to begin with, because people would just compress everything in different sites/apps/CLIs

**Comment by pornel** [comment]
The lossless vs lossy is a tough one. 

People define "lossless" differently. For me it's "doesn't change pixel values at all", but for many people it's just "can't see the difference with naked eye". 

This issue is not unusual — I keep hearing from users that they find some other tool that makes files smaller, and almost always it turns out it's only because the other tool makes quality (slightly) lower.

So instead of explaining over and over again that there's technically a gigantic difference between strictly lossless and lossy "can't see loss", I've just added lossy option so I can tell people to use it and get the file size reduction they want.

**Comment by Saklad5** [comment]
I don’t think the confusion about lossless meaning “not very lossy” is common enough to be an acceptable use. As I define it, lossless means literally no data loss, and all lossless processes have to be theoretically reversible.

**Comment by pornel** [comment]
It affects more users than just those opening issue on GitHub. Explaining users that they're wrong is not as good as providing them what they want.

Other apps have lossy options they call "without compromising quality" or even "visually lossless". I think I'm more honest with this calling it lossy and making it opt-in.

**Comment by Saklad5** [comment]
I agree. I'm just saying you shouldn't feel pressured to compete with tools that are intended specifically for lossy compression like that. Unless you want to, of course.
